sdk/testing/src/main/java/io/opentelemetry/sdk/testing/junit4/OpenTelemetryRule.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.testing.junit4;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.trace.propagation.HttpTraceContext;
+import io.opentelemetry.context.propagation.DefaultContextPropagators;
+import io.opentelemetry.exporter.inmemory.InMemorySpanExporter;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.trace.TracerSdkManagement;
+import io.opentelemetry.sdk.trace.TracerSdkProvider;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor;
+import java.util.List;
+import org.junit.rules.ExternalResource;
+
+/**
+ * A JUnit4 rule which sets up the {@link OpenTelemetrySdk} for testing, resetting state between
+ * tests. This rule cannot be used with {@link org.junit.ClassRule}.
+ *
+ * <pre>{@code
+ * > public class CoolTest {
+ * >   {@literal @}Rule
+ * >   public OpenTelemetryExtension otelTesting = OpenTelemetryExtension.create();
+ * >
+ * >   private Tracer tracer;
+ * >
+ * >   {@literal @}Before
+ * >   public void setUp() {
+ * >       tracer = otelTesting.getOpenTelemetry().getTracer("test");
+ * >   }
+ * >
+ * >   {@literal @}Test
+ * >   public void test() {
+ * >     tracer.spanBuilder("name").startSpan().end();
+ * >     assertThat(otelTesting.getSpans()).containsExactly(expected);
+ * >   }
+ * >  }
+ * }</pre>
+ */
+public class OpenTelemetryRule extends ExternalResource {
+
+  /**
+   * Returns a {@link OpenTelemetryRule} with a default SDK initialized with an in-memory span
+   * exporter and W3C trace context propagation.
+   */
+  public static OpenTelemetryRule create() {
+    InMemorySpanExporter spanExporter = InMemorySpanExporter.create();
+
+    TracerSdkProvider tracerProvider = TracerSdkProvider.builder().build();
+    tracerProvider.addSpanProcessor(SimpleSpanProcessor.builder(spanExporter).build());
+
+    OpenTelemetrySdk openTelemetry =
+        OpenTelemetrySdk.builder()
+            .setPropagators(
+                DefaultContextPropagators.builder()
+                    .addTextMapPropagator(HttpTraceContext.getInstance())
+                    .build())
+            .setTracerProvider(tracerProvider)
+            .build();
+
+    return new OpenTelemetryRule(openTelemetry, spanExporter);
+  }
+
+  private final OpenTelemetrySdk openTelemetry;
+  private final InMemorySpanExporter spanExporter;
+
+  private OpenTelemetry previousGlobalOpenTelemetry;
```

Comment:
does this need to be `volatile` ? I'm never 100% clear on the threading model of junit.

Reply:
Good point - I'm not sure either and it doesn't hurt so added